pkg/minishift/generic/generic.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generic
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/docker/machine/libmachine/provision"
+	"github.com/minishift/minishift/pkg/minikube/sshutil"
+	"golang.org/x/crypto/ssh"
+)
+
+func PrepareRemoteMachine(s *ssh.Client) error {
+	osReleaseOut, err := detectOS(s)
+	if err != nil {
+		return err
+	}
+	osReleaseInfo, err := provision.NewOsRelease([]byte(osReleaseOut))
+	if err != nil {
+		return err
+	}
+	if osReleaseInfo.ID == "fedora" || osReleaseInfo.ID == "rhel" || osReleaseInfo.ID == "centos" {
+		if err := prepareRHELVariant(s); err != nil {
+			return err
```

Comment:
We are calling `PrepareRemoteMachine` before the generic driver kicks in. So if anyone want to add similar support of other operating systems they might need to add code to `PrepareRemoteMachine`. Though it depends on what libmachine has for that os. IMO we should implement an interface here and each OS should have `setupHostForGenericDriver` (or something similar) function which should have actual automation.  WDYT?

Reply:
>Though it depends on what libmachine has for that os. IMO we should implement an interface here and each OS should have setupHostForGenericDriver (or something similar) function which should have actual automation. WDYT?

@LalatenduMohanty that is already part of each provisioner here we are explicit doing because libmachine provisioner for RHEL/CentOS and Fedora is outdated once we start consuming the machine-driver/libmachine than we might not need it as this can become part of standard provisioner. I don't want to boated lot of stuff which we may need in future. (If we need we can always refactor). Remember we invested lot of time for registration thinking that might be Suse can implement their own way of subscription but here we are code is there but no outside contribution. Everything shouldn't be solved by interface (it good concept but using it to predict that something can be used in future doesn't make sense to me).

Reply:
@praveenkumar  Having a flexible code always helps in long term. If you want to do this step by step I am ok with talking this change and we can refactor later.

Reply:
If you look for flexible code, we need a pre-provision `add-on-like` executioner. This will run to deploy packages or setups from files. This is what I see as a refactor...

Reply:
@gbraad when you say `add-on-like` executioner you mean have a plain file which user can specify and contain the commands which should run before hand?

Reply:
yep... not Ansible, and not a real sh-bang (#!/bin/sh) shell file...

A file that can be specified to run on a particular target based on meta-data perhaps. But this is food for thought, not to do now...